### ISSUE TITLE
Resolving remote URIs in Resources

### ISSUE THREAD

**Issue by ewoudwerkman** [pyecore/pyecore#69]
EMF and PyEcore have the ability to load additional resources in a ResourceSet, but resolving resources that have a remote URI (```http://...```) does not yet work in PyEcore.

I've created a patch to overcome this, but it feels a little like a hack. See the attached Pull request.
What I do is checking if the URI-string starts with ```http``` and then wrap it in a ```HttpURI``` instead of a ```URI```. This was needed because a Resource in pyEcore wraps everything in a URI and does not much with the protocol part of the URI.

What would be better in my opinion is to merge URI and HttpURI and have different behavior based on the protocol scheme (e.g. no protocol or file:// protocol -> local file, use open, http:// and https:// use urllib, etc.)

In EMF this is done by the URIConverter that each Resource has. This URIConverter has a (static, extendible ) list of URI handlers, e.g. one for the eclipse platform:// URI, and for zip:// and jar:// archived-based URIs and for http(s):// based URIs. Each URIHandler has a method ```canHandle()``` that returns true if it can handle the URI. The first handler in the list that returns true is used. 

I did not implement this, since I did not dive into the current URI details (e.g. checking if the URI has the same root (apply_relative, etc.))
What are your thoughts?

**Comment by ewoudwerkman** [comment]
See #70 for the patch

**Comment by aranega** [comment]
Hi @ewoudwerkman !

Thanks a lot for the suggestion, you're right, the current design is not ideal at all at is doesn't allow you to select at runtime the kind of URI you want. Your issue is closely related to what @CFAndy asked for in [#33](https://github.com/pyecore/pyecore/issues/33#issuecomment-435762304) .

With your idea, I created a first version that tackle that. The version is not yet ideal and still requires tests, but it would work in this way (you can find the modified version in the `feature/customize_uri_creation` branch), this example is for an `myecore.ecore` metamodel with a reference to an exeternal remote resource in the form `https://....`:

```python
from pyecore.resources.resource import HttpURIConverter

rset = ResourceSet()
rset.uri_converter.append(HttpURIConverter)
resource = rset.get_resource('myecore.ecore')
root = resource.contents[0]
```

The code of the `HttpURIConverter` is the following:

```python
class HttpURIConverter(object):
    @staticmethod
    def can_handle(uri):
        return uri.protocol == 'http' or uri.protocol == 'https'

    @staticmethod
    def convert(uri):
        return HttpURI(uri.plain)
```

Would this solution be acceptable for you?

**Comment by ewoudwerkman** [comment]
Sure! This seems like a good and scalable solution (i.e. fits well for other protocol schemes, too!). 
The only question I have is if the http(s) protocol should be supported out of the box for pyEcore, instead of configuring it?

**Comment by aranega** [comment]
That's really a good question, to be honest, I really don't know. EMF doesn't propose it by default, so the most logical thing would be to not place it, also because it could imply most complex debug sessions. But one of the goal of pyecore was to make EMF a little bit easier (especially the dynamic part), so ... It would be better to have it already installed.

I think I would add it by default, perhaps someone has another reason why it should not?

1. I need to perform extra tests before including it, just to be sure this doesn't imply interferences with the `URIMapper` behavior.

**Comment by ewoudwerkman** [comment]
The adaptations look great and more generic applicable.
I added an abstract class that can be subclassed by user implentations, and added it to the ```global_uri_converter``` in ```__init()__.py```. Feel free to adapt. See #71 

I don't really get what URITranslator does and it is a little confusing with URIConverter. Could you come up with a different name for that? Its method name should also reflect its behavior.

**Comment by aranega** [comment]
Thanks for the PR, I'll integrate it right away.

The `URITranslator` is actually here for the URI mapping logic, I renamed it quickly from it's previous name which was too confusing with the `URIConverter` logic. I'll rename it into `URIMapper` and `convert` to `translate`.

**Comment by ewoudwerkman** [comment]
That makes sense. Perfect!
Would you mind to issue a new release with these changes?

**Comment by aranega** [comment]
No problem, I just need to push tests about this in order to be sure that everything is ok (as far as I saw, no test is breaking, but more would be better). I still have a problem with the CI lately which doesn't like the fact that I override `isinstance` on some very specific Python versions. I need to try to find a way to fix this also. I will try to make a new release during the end of the week or this weekend.

**Comment by ewoudwerkman** [comment]
Thanks!
Yes I saw that CI failed on older python versions, but have no clue what was happening there.

**Comment by aranega** [comment]
Strangely on some specific Python versions (3.5.6 for example, but not 3.5.4), the fact that I override a buitin is not well accepted by the features related to generics in the types annotation module. I will try to find a solution for that, it's really bothering me... The main point of overriding the `issubclass` function was to ease the implementation of the profiles parts for the pyuml2 project. If I don't find any solution, I will migrate this patch to pyuml2 instead (I hope this functionnality it is not used in other project...)

**Comment by aranega** [comment]
@ewoudwerkman I released a new version of PyEcore with the patch included, you can update to version `0.11.0` :).
